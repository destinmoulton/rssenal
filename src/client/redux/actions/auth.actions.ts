import { API_AUTH } from "../apiendpoints";
import { JSON_HEADERS } from "../../lib/headers";

import { IDispatch } from "../../interfaces";

export function authenticateUser(username: string, password: string) {
    return (dispatch: IDispatch) => {
        const url = API_AUTH;
        const init = {
            method: "POST",
            body: JSON.stringify({ username, password }),
            headers: JSON_HEADERS
        };

        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(auth => {
                if (auth.isAuthorized) {
                }
            })
            .catch(err => {});
    };
}
