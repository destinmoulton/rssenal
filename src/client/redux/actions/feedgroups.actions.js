import {
    FEEDGROUPS_FETCHING,
    FEEDGROUPS_RECEIVED,
    FEEDGROUPS_UPDATE_BEGIN,
    FEEDGROUPS_UPDATE_COMPLETE
} from "../actiontypes";

import {
    API_FEEDGROUPS_BASE,
    API_FEEDGROUPS_GET_ALL
} from "../apiendpoints";

export function getAllFeedGroups(){
    return (dispatch)=>{
        dispatch(fetchingInProgress());
        dispatch(fetchFeedGroups());
    }
}

function fetchingInProgress(){
    return {
        type: FEEDGROUPS_FETCHING
    }
}

function fetchFeedGroups(){
    return (dispatch)=>{
        const url = API_FEEDGROUPS_GET_ALL;
        const init = {
            method: "GET"
        };
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((groups)=>{
                dispatch(fetchingComplete(groups));
            })
    }
}

function fetchingComplete(data){
    return {
        type: FEEDGROUPS_RECEIVED,
        groups: data
    }
}

export function beginSaveFeedGroup(groupId, newGroupName){
    return (dispatch)=>{
        dispatch(updatingInProgress(groupId));
        dispatch(updateFeedGroup(groupId, newGroupName));
    };
}

function updatingInProgress(groupId){
    return {
        type: FEEDGROUPS_UPDATE_BEGIN,
        groupId
    };
}

function updateFeedGroup(groupId, newGroupName){
    return (dispatch)=>{
        const url = API_FEEDGROUPS_BASE + groupId;
        const init = {
            method: "PUT",
            body: JSON.stringify({name: newGroupName}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((newGroup)=>{
                dispatch(updatingComplete(newGroup))
            })
    }
}

function updatingComplete(newGroup){
    return {
        type: FEEDGROUPS_UPDATE_COMPLETE,
        group: newGroup
    }
}