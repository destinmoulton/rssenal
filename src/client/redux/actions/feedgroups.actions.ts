import {
    FEEDGROUPS_FETCHING,
    FEEDGROUPS_RECEIVED,
    FEEDGROUPS_ADD_BEGIN,
    FEEDGROUPS_ADD_COMPLETE,
    FEEDGROUPS_DELETE_BEGIN,
    FEEDGROUPS_DELETE_COMPLETE,
    FEEDGROUPS_UPDATE_BEGIN,
    FEEDGROUPS_UPDATE_COMPLETE
} from "../actiontypes";

import {
    API_FEEDGROUPS_BASE,
    API_FEEDGROUPS_GET_ALL
} from "../apiendpoints";

import { IDispatch, IFeedgroup, TFeedgroupID } from "../../interfaces";

export function getAllFeedGroups(){
    return (dispatch: IDispatch)=>{
        dispatch(fetchingInProgress());
        dispatch(fetchFeedGroups());
    }
}

function fetchingInProgress(){
    return {
        type: FEEDGROUPS_FETCHING
    }
}

function fetchFeedGroups(){
    return (dispatch: IDispatch)=>{
        const url = API_FEEDGROUPS_GET_ALL;
        const init = {
            method: "GET"
        };
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((groups)=>{
                dispatch(fetchingComplete(groups));
            })
    }
}

function fetchingComplete(data: IFeedgroup[]){
    return {
        type: FEEDGROUPS_RECEIVED,
        groups: data
    }
}

export function beginSaveFeedGroup(groupInfo: IFeedgroup){
    return (dispatch: IDispatch)=>{
        if(groupInfo._id !== ""){
            dispatch(updatingInProgress());
            dispatch(updateFeedGroup(groupInfo._id, groupInfo.name));
        } else {
            dispatch(addingInProgress());
            dispatch(addFeedGroup(groupInfo.name));
        }
    }
}

function addingInProgress(){
    return {
        type: FEEDGROUPS_ADD_BEGIN
    }
}

function addFeedGroup(newGroupName: string){
    return (dispatch: IDispatch)=>{
        const url = API_FEEDGROUPS_BASE;
        const init = {
            method: "POST",
            body: JSON.stringify({name: newGroupName}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((newGroup)=>{
                dispatch(addingFeedGroupComplete(newGroup));
            })
    }
}

function addingFeedGroupComplete(group: IFeedgroup){
    return {
        type: FEEDGROUPS_ADD_COMPLETE,
        group
    };
}

function updatingInProgress(){
    return {
        type: FEEDGROUPS_UPDATE_BEGIN
    };
}

function updateFeedGroup(groupId: TFeedgroupID, newGroupName: string){
    return (dispatch: IDispatch)=>{
        const url = API_FEEDGROUPS_BASE + groupId;
        const init = {
            method: "PUT",
            body: JSON.stringify({name: newGroupName}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((newGroup)=>{
                dispatch(updatingComplete(newGroup))
            })
    }
}

function updatingComplete(newGroup: IFeedgroup){
    return {
        type: FEEDGROUPS_UPDATE_COMPLETE,
        group: newGroup
    }
}

export function beginDeleteFeedGroup(groupId: TFeedgroupID){
    return (dispatch: IDispatch)=>{
        dispatch(deleteInProgress());
        dispatch(deleteFeedGroup(groupId));
    }
}

function deleteInProgress(){
    return {
        type: FEEDGROUPS_DELETE_BEGIN
    }
}

function deleteFeedGroup(groupId: TFeedgroupID){
    return (dispatch: IDispatch)=>{
        const url = API_FEEDGROUPS_BASE + groupId;
        const init = {
            method: "DELETE"
        }
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((resObj)=>{
                if(resObj.status==="success"){
                    dispatch(deleteComplete(groupId));
                }
            })
    }
}

function deleteComplete(groupId: TFeedgroupID){
    return {
        type: FEEDGROUPS_DELETE_COMPLETE,
        groupId
    }
}