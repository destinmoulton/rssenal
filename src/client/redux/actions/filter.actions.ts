import { OrderedMap } from "immutable";

import { FILTER_CHANGE, FILTER_RESET } from "../actiontypes";
import * as Types from "../../types";

export function filterChangeActive(filter: Types.IFilter) {
    return (dispatch: Types.IDispatch, getState: Types.IGetState) => {
        const { entriesStore } = getState();
        dispatch(filterVisibleEntries(filter, entriesStore.entries));
    };
}

export function filterVisibleEntries(
    filter: Types.IFilter,
    currentEntries: Types.TEntries
) {
    return (dispatch: Types.IDispatch, getState: Types.IGetState) => {
        const { feedsStore, foldersStore } = getState();
        const allFeeds = feedsStore.feeds;

        let title = "All";

        let filteredEntries = currentEntries.toOrderedMap();
        switch (filter.limit) {
            case "feed":
                const activeFeed = feedsStore.feeds.get(filter.id);
                title = activeFeed.title;

                filteredEntries = currentEntries
                    .filter(entry => {
                        return entry.feed_id === filter.id;
                    })
                    .toOrderedMap();
                break;
            case "folder":
                if (filter.id !== "all") {
                    const activeFolder = foldersStore.folders.get(filter.id);

                    title = activeFolder.name;

                    const feedIds = allFeeds
                        .filter(feed => {
                            return feed.folder_id === filter.id;
                        })
                        .map(feed => {
                            return feed._id;
                        });

                    filteredEntries = currentEntries
                        .filter(entry => {
                            return feedIds.includes(entry.feed_id);
                        })
                        .toOrderedMap();
                }
                break;
        }
        dispatch(setNewFilter(filter, title, filteredEntries));
    };
}

function setNewFilter(
    newFilter: Types.IFilter,
    filterTitle: string,
    filteredEntries: OrderedMap<Types.TEntryID, Types.IEntry>
) {
    return {
        type: FILTER_CHANGE,
        newFilter,
        filterTitle,
        filteredEntries
    };
}

export function filterReset() {
    return {
        type: FILTER_RESET
    };
}
