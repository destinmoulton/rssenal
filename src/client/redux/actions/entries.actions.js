
import {
    ENTRIES_GET_BEGIN,
    ENTRIES_GET_COMPLETE,
    ENTRIES_UPDATE_BEGIN,
    ENTRIES_UPDATE_COMPLETE
} from "../actiontypes";

import {
    API_ENTRIES_BASE
} from "../apiendpoints";

import { feedsDecrementUnread, feedsSetAllUnreadCount } from "./feeds.actions";

export function beginGetEntries(){
    return (dispatch)=>{
        dispatch(beginGetProcess());
        dispatch(getEntries());
    }
}

function beginGetProcess(){
    return {
        type: ENTRIES_GET_BEGIN
    }
}

function getEntries(){
    return (dispatch, getState)=>{
        const url = API_ENTRIES_BASE;
        const init = {
            method: "GET"
        };

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((resObj)=>{
                if(resObj.status === "error"){
                    console.error(resObj.error);
                } else {
                    let entries = resObj.entries;
                    dispatch(getEntriesComplete(entries));
                    dispatch(entriesUpdateUnreadCount(entries));
                }
            })
    }
}

function getEntriesComplete(entries){
    return {
        type: ENTRIES_GET_COMPLETE,
        entries
    }
}

function entriesUpdateUnreadCount(entries){
    return (dispatch)=>{
        dispatch(feedsSetAllUnreadCount(entries));
    }
}

export function updateReadState(entry, hasRead){
    return (dispatch)=>{
        const url = API_ENTRIES_BASE + entry._id;
        const init = {
            method: "PUT",
            body: JSON.stringify({has_read: hasRead}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((resObj)=>{
                if(resObj.status === "error"){
                    console.error(resObj.error);
                } else {
                    dispatch(feedsDecrementUnread(entry.feed_id));
                }
            });
    }
}