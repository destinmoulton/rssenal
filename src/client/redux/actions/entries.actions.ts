import { OrderedMap } from "immutable";
import * as moment from "moment";

import { ENTRIES_CLEAR_ALL, ENTRIES_SET_ALL } from "../actiontypes";

import { API_ENTRIES_BASE } from "../apiendpoints";

import { generateJWTJSONHeaders, generateJWTHeaders } from "../../lib/headers";
import { feedsDecrementUnread, feedsSetAllUnreadCount } from "./feeds.actions";
import { filterVisibleEntries } from "./filter.actions";
import { message } from "./messages.actions";

import {
    IDispatch,
    IEntry,
    IGetState,
    TFeedID,
    TEntryID,
    TEntries
} from "../../interfaces";

export function getEntriesForFeed(feedId: TFeedID) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { settingsStore } = getState();
        let showRead = false;
        settingsStore.settings.forEach(setting => {
            if (setting.key === "show_entries_has_read") {
                showRead = setting.value;
            }
        });

        const queryString =
            "?showEntriesHasRead=" + showRead + "&feedId=" + feedId;
        const url = API_ENTRIES_BASE + queryString;
        const init = {
            method: "GET",
            headers: generateJWTHeaders()
        };

        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(resObj => {
                if (resObj.status === "error") {
                    dispatch(message(resObj.error, "error"));
                    console.error(resObj.error);
                } else {
                    let entries = resObj.entries;
                    dispatch(ammendEntries(entries));
                }
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function ammendEntries(entries: IEntry[]) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { feedsStore, filterStore } = getState();

        const ammendedEntries = entries.map((entry: IEntry) => {
            const feedTitle = feedsStore.feeds.get(entry.feed_id).title;
            const timeAgo = moment(entry.publish_date).fromNow();
            return {
                ...entry,
                feedTitle,
                timeAgo
            };
        });
        dispatch(getEntriesComplete(ammendedEntries));

        dispatch(feedsSetAllUnreadCount(ammendedEntries));
    };
}

function getEntriesComplete(entries: IEntry[]) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { entriesStore, filterStore } = getState();
        const currentEntries = entriesStore.entries;

        let newEntries = currentEntries.toOrderedMap();
        entries.forEach(entry => {
            newEntries = newEntries.set(entry._id, entry);
        });

        dispatch(entriesSetAndFilter(newEntries));
    };
}

export function updateReadState(entry: IEntry, hasRead: boolean) {
    return (dispatch: IDispatch) => {
        const url = API_ENTRIES_BASE + entry._id;
        const init = {
            method: "PUT",
            body: JSON.stringify({ has_read: hasRead }),
            headers: generateJWTJSONHeaders()
        };

        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(resObj => {
                if (resObj.status === "error") {
                    console.error(resObj.error);
                } else {
                    dispatch(entryAmmendMarkRead(entry._id));
                }
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function entryAmmendMarkRead(entryId: TEntryID) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { entriesStore, filterStore } = getState();
        const allEntries = entriesStore.entries;
        const entry = allEntries.get(entryId);

        const newEntry = { ...entry, has_read: true };
        const newEntries = allEntries.set(entryId, newEntry);

        dispatch(feedsDecrementUnread(entry.feed_id));
        dispatch(entriesSetAndFilter(newEntries));
    };
}

export function entriesRemoveFeed(feedId: TFeedID) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { entriesStore } = getState();
        const { entries } = entriesStore;

        const newEntries = entries
            .filter((entry: IEntry) => {
                return entry.feed_id !== feedId;
            })
            .toOrderedMap();
        dispatch(entriesSetAndFilter(newEntries));
    };
}

function entriesSetAndFilter(entries: TEntries) {
    return (dispatch: IDispatch, getState: IGetState) => {
        const { filterStore } = getState();
        dispatch(entriesSetAll(entries));
        dispatch(filterVisibleEntries(filterStore.filter, entries));
    };
}

function entriesSetAll(entries: TEntries) {
    return {
        type: ENTRIES_SET_ALL,
        entries
    };
}

export function entriesClearAll() {
    return {
        type: ENTRIES_CLEAR_ALL
    };
}
