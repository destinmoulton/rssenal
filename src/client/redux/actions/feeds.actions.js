
import {
    FEEDS_ADD_BEGIN,
    FEEDS_ADD_COMPLETE,
    FEEDS_DECREMENT_UNREAD,
    FEEDS_DELETE_BEGIN,
    FEEDS_DELETE_COMPLETE,
    FEEDS_GETALL_COMPLETE,
    FEEDS_SETALL_UNREAD_COUNT,
    FEEDS_UPDATE_BEGIN,
    FEEDS_UPDATE_COMPLETE
} from "../actiontypes";

import {
    API_FEEDS_BASE
} from "../apiendpoints";

import {beginUpdateAndGetEntries} from "./entries.actions";

export function beginAddFeed(feedInfo){
    return (dispatch)=>{
        dispatch(beginAddFeedProcess());
        dispatch(addFeed(feedInfo));
    }
}

function beginAddFeedProcess(){
    return {
        type: FEEDS_ADD_BEGIN
    }
}

function addFeed(feedInfo){
    return (dispatch)=>{
        const url = API_FEEDS_BASE;
        const init = {
            method: "POST",
            body: JSON.stringify({...feedInfo}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((feedObj)=>{
                if(feedObj.status === "error"){
                    console.error(feedObj.error);
                } else if(feedObj.status === "success"){
                    dispatch(addFeedComplete(feedObj.feedInfo));
                    dispatch(beginUpdateAndGetEntries());
                }
            })
    }
}

function addFeedComplete(feed){
    return {
        type: FEEDS_ADD_COMPLETE,
        feed
    }
}

export function getAllFeeds(){
    return (dispatch)=>{
        const url = API_FEEDS_BASE;
        const init = {
            method: "GET"
        };
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((resObj)=>{
                if(resObj.status==="error"){
                    console.error(resObj.error);
                } else if(resObj.status==="success"){
                    dispatch(getAllFeedsComplete(resObj.feeds));
                }
            })
    }
}

function getAllFeedsComplete(feeds){
    return {
        type: FEEDS_GETALL_COMPLETE,
        feeds
    }
}

export function beginDeleteFeed(feedId){
    return (dispatch)=>{
        dispatch(beginDeleteProcess());
        dispatch(deleteFeed(feedId));
    }
}

function beginDeleteProcess(feedId){
    return {
        type: FEEDS_DELETE_BEGIN
    }
}

function deleteFeed(feedId){
    return (dispatch)=>{
        const url = API_FEEDS_BASE + feedId;
        const init = {
            method: "DELETE"
        }
        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((resObj)=>{
                dispatch(deleteFeedComplete(feedId));
                dispatch(beginUpdateAndGetEntries());
            })
    }
}

function deleteFeedComplete(feedId){
    return {
        type: FEEDS_DELETE_COMPLETE,
        feedId
    }
}

export function beginUpdateFeed(feedInfo){
    return (dispatch)=>{
        dispatch(beginUpdateFeedProcess());
        dispatch(updateFeed(feedInfo));
    }
}

function beginUpdateFeedProcess(){
    return {
        type: FEEDS_UPDATE_BEGIN
    }
}

function updateFeed(feedInfo){
    return (dispatch)=>{
        const url = API_FEEDS_BASE + feedInfo._id;
        const init = {
            method: "PUT",
            body: JSON.stringify({title: feedInfo.title, feedgroup_id: feedInfo.feedgroup_id}),
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }

        fetch(url, init)
            .then((res)=>{
                return res.json();
            })
            .then((feedObj)=>{
                if(feedObj.status === "error"){
                    console.error(feedObj.error);
                } else if(feedObj.status === "success"){
                    dispatch(updateFeedComplete(feedObj.feedInfo));
                }
            })
    }
}

function updateFeedComplete(feed){
    return {
        type: FEEDS_UPDATE_COMPLETE,
        feed
    }
}

export function feedsSetAllUnreadCount(entries){
    return {
        type: FEEDS_SETALL_UNREAD_COUNT,
        entries
    }
}

export function feedsDecrementUnread(feedId){
    return {
        type: FEEDS_DECREMENT_UNREAD,
        feedId
    }
}