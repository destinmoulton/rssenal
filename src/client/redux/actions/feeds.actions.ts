import {
    FEEDS_ADD_BEGIN,
    FEEDS_ADD_COMPLETE,
    FEEDS_DECREMENT_UNREAD,
    FEEDS_DELETE_BEGIN,
    FEEDS_DELETE_COMPLETE,
    FEEDS_GETALL_COMPLETE,
    FEEDS_CALC_UNREAD_COUNT,
    FEEDS_CLEAR_UNREAD,
    FEEDS_UPDATE_BEGIN,
    FEEDS_UPDATE_COMPLETE
} from "../actiontypes";

import { API_FEEDS_BASE } from "../apiendpoints";

import { generateJWTJSONHeaders, generateJWTHeaders } from "../../lib/headers";

import { getEntriesForFeed, entriesRemoveFeed } from "./entries.actions";
import { message } from "./messages.actions";

import {
    IDispatch,
    IFeed,
    TFeedID,
    TEntries,
    IRootStoreState
} from "../../interfaces";

export function beginAddFeed(feedInfo: IFeed) {
    return (dispatch: IDispatch) => {
        dispatch(beginAddFeedProcess());
        dispatch(addFeed(feedInfo));
    };
}

function beginAddFeedProcess() {
    return {
        type: FEEDS_ADD_BEGIN
    };
}

function addFeed(feedInfo: IFeed) {
    return (dispatch: IDispatch) => {
        const url = API_FEEDS_BASE;
        const headers = generateJWTJSONHeaders();
        const init = {
            method: "POST",
            body: JSON.stringify({ ...feedInfo }),
            headers
        };

        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(feedObj => {
                if (feedObj.status === "error") {
                    console.error(feedObj.error);
                } else if (feedObj.status === "success") {
                    dispatch(message("Feed added.", "success"));
                    dispatch(addFeedComplete(feedObj.feedInfo));
                    dispatch(getEntriesForFeed(feedObj.feedInfo._id));
                }
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function addFeedComplete(feed: IFeed) {
    return {
        type: FEEDS_ADD_COMPLETE,
        feed
    };
}

export function getAllFeeds() {
    return (dispatch: IDispatch) => {
        const url = API_FEEDS_BASE;
        const init = {
            method: "GET",
            headers: generateJWTHeaders()
        };
        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(resObj => {
                if (resObj.status === "error") {
                    console.error(resObj.error);
                } else if (resObj.status === "success") {
                    dispatch(getAllFeedsComplete(resObj.feeds));
                    dispatch(getAllEntriesForFeeds(resObj.feeds));
                }
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function getAllFeedsComplete(feeds: IFeed[]) {
    return {
        type: FEEDS_GETALL_COMPLETE,
        feeds
    };
}

function getAllEntriesForFeeds(feeds: IFeed[]) {
    return (dispatch: IDispatch) => {
        dispatch(clearUnread());
        feeds.forEach(feed => {
            dispatch(getEntriesForFeed(feed._id));
        });
    };
}

function clearUnread() {
    return {
        type: FEEDS_CLEAR_UNREAD
    };
}

export function refreshAllFeeds() {
    return (dispatch: IDispatch, getState: () => IRootStoreState) => {
        const { feeds } = getState().feeds;
        dispatch(getAllEntriesForFeeds(feeds.toArray()));
    };
}

export function beginDeleteFeed(feedId: TFeedID) {
    return (dispatch: IDispatch) => {
        dispatch(beginDeleteProcess());
        dispatch(deleteFeed(feedId));
    };
}

function beginDeleteProcess() {
    return {
        type: FEEDS_DELETE_BEGIN
    };
}

function deleteFeed(feedId: TFeedID) {
    return (dispatch: IDispatch) => {
        const url = API_FEEDS_BASE + feedId;
        const init = {
            method: "DELETE",
            headers: generateJWTHeaders()
        };
        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(resObj => {
                dispatch(message("Feed removed.", "success"));
                dispatch(deleteFeedComplete(feedId));
                dispatch(entriesRemoveFeed(feedId));
                dispatch(getAllFeeds());
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function deleteFeedComplete(feedId: TFeedID) {
    return {
        type: FEEDS_DELETE_COMPLETE,
        feedId
    };
}

export function beginUpdateFeed(feedInfo: IFeed) {
    return (dispatch: IDispatch) => {
        dispatch(beginUpdateFeedProcess());
        dispatch(updateFeed(feedInfo));
    };
}

function beginUpdateFeedProcess() {
    return {
        type: FEEDS_UPDATE_BEGIN
    };
}

function updateFeed(feedInfo: IFeed) {
    return (dispatch: IDispatch, getState: () => IRootStoreState) => {
        const url = API_FEEDS_BASE + feedInfo._id;
        const init = {
            method: "PUT",
            body: JSON.stringify({
                title: feedInfo.title,
                folder_id: feedInfo.folder_id
            }),
            headers: generateJWTJSONHeaders()
        };

        fetch(url, init)
            .then(res => {
                return res.json();
            })
            .then(feedObj => {
                if (feedObj.status === "error") {
                    console.error(feedObj.error);
                } else if (feedObj.status === "success") {
                    dispatch(message("Feed saved.", "success"));
                    dispatch(updateFeedComplete(feedObj.feedInfo));
                    dispatch(getAllFeeds());
                }
            })
            .catch(err => {
                console.error(err);
            });
    };
}

function updateFeedComplete(feed: IFeed) {
    return {
        type: FEEDS_UPDATE_COMPLETE,
        feed
    };
}

export function feedsSetAllUnreadCount(entries: TEntries) {
    return {
        type: FEEDS_CALC_UNREAD_COUNT,
        entries
    };
}

export function feedsDecrementUnread(feedId: TFeedID) {
    return {
        type: FEEDS_DECREMENT_UNREAD,
        feedId
    };
}
