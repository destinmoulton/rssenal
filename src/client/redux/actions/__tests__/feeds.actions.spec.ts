import { List, OrderedMap } from "immutable";
import * as fetchMock from "fetch-mock";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

import * as FeedsActions from "../feeds.actions";
import * as ACT_TYPES from "../../actiontypes";
import * as INIT_STATE from "../../initialstate";

import API_ENTRIES_STRING from "../../../../../test/data/api/entries.unread";
import API_FEEDS_STRING from "../../../../../test/data/api/feeds";
import FEED from "../../../../../test/data/feed";
import AMMENDED_ENTRIES from "../../../../../test/data/immutable/ammendedEntries";
import IMM_FEEDS from "../../../../../test/data/immutable/feeds";
import * as IMM_UNREAD from "../../../../../test/data/immutable/unreadMap";

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

describe("feeds.actions", () => {
    afterEach(() => {
        fetchMock.reset();
        fetchMock.restore();
        localStorage.clear();
    });

    it("feedAdd() handles adding a feed", () => {
        const feedID = "5b33c76cb2438d5708dc197e";
        const entries_url =
            "/api/entries/?showEntriesHasRead=false&feedId=" + feedID;

        fetchMock.getOnce(entries_url, JSON.parse(API_ENTRIES_STRING));

        const feed_url = "/api/feeds/";

        fetchMock.postOnce(feed_url, { status: "success", feedInfo: FEED });

        const store = mockStore({
            entriesStore: INIT_STATE.ENTRIES_INITIAL_STATE,
            feedsStore: {
                ...INIT_STATE.FEEDS_INITIAL_STATE,
                feeds: IMM_FEEDS
            },
            filterStore: INIT_STATE.FILTER_INITIAL_STATE,
            foldersStore: INIT_STATE.FOLDERS_INITIAL_STATE,
            messagesStore: INIT_STATE.MESSAGES_INITIAL_STATE,
            settingsStore: INIT_STATE.SETTINGS_INITIAL_STATE
        });

        const newFeed = {
            title: "NEW_FEED",
            folder_id: "FOLDER_ID",
            description: "DESCRIPTION",
            link: "LINK_HERE"
        };

        return store.dispatch(FeedsActions.feedAdd(newFeed)).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        });
    });

    it("feedsGetAll() handles getting feeds", () => {
        const feedID = "5b33c76cb2438d5708dc197e";
        const entries_url =
            "/api/entries/?showEntriesHasRead=false&feedId=" + feedID;

        fetchMock.getOnce(entries_url, JSON.parse(API_ENTRIES_STRING));

        const feed_url = "/api/feeds/";

        fetchMock.getOnce(feed_url, JSON.parse(API_FEEDS_STRING));

        const store = mockStore({
            entriesStore: INIT_STATE.ENTRIES_INITIAL_STATE,
            feedsStore: {
                ...INIT_STATE.FEEDS_INITIAL_STATE,
                feeds: IMM_FEEDS
            },
            filterStore: INIT_STATE.FILTER_INITIAL_STATE,
            foldersStore: INIT_STATE.FOLDERS_INITIAL_STATE,
            messagesStore: INIT_STATE.MESSAGES_INITIAL_STATE,
            settingsStore: INIT_STATE.SETTINGS_INITIAL_STATE
        });

        return store.dispatch(FeedsActions.feedsGetAll()).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        });
    });

    it("feedsRefreshAll() handles refreshing entries for current feeds", () => {
        const feedID = "5b33c76cb2438d5708dc197e";
        const entries_url =
            "/api/entries/?showEntriesHasRead=false&feedId=" + feedID;

        fetchMock.getOnce(entries_url, JSON.parse(API_ENTRIES_STRING));

        const store = mockStore({
            entriesStore: INIT_STATE.ENTRIES_INITIAL_STATE,
            feedsStore: {
                ...INIT_STATE.FEEDS_INITIAL_STATE,
                feeds: IMM_FEEDS
            },
            filterStore: INIT_STATE.FILTER_INITIAL_STATE,
            foldersStore: INIT_STATE.FOLDERS_INITIAL_STATE,
            messagesStore: INIT_STATE.MESSAGES_INITIAL_STATE,
            settingsStore: INIT_STATE.SETTINGS_INITIAL_STATE
        });

        store.dispatch(FeedsActions.feedsRefreshAll());
        expect(store.getActions()).toMatchSnapshot();
    });
});
