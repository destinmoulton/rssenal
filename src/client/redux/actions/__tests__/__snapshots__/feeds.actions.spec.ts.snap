// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feeds.actions feedAdd() handles adding a feed 1`] = `
Array [
  Object {
    "type": "FEEDS_ADD_BEGIN",
  },
  Object {
    "lastUID": 1,
    "messages": Immutable.List [
      Object {
        "level": "success",
        "message": "Feed added.",
        "uid": 1,
      },
    ],
    "type": "MESSAGES_ADD_COMPLETE",
  },
  Object {
    "feeds": Immutable.OrderedMap {
      "5b33c76cb2438d5708dc197e": Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc197e",
        "creation_date": "2018-06-27T17:20:44.037Z",
        "description": null,
        "folder_id": "5b33c57eb2438d5708dc1968",
        "link": "http://highscalability.com/blog/",
        "title": "High Scalability",
        "url": "http://feeds.feedburner.com/HighScalability",
      },
    },
    "type": "FEEDS_ADD_COMPLETE",
  },
  Object {
    "type": "ENTRIES_GET_ALL_REQUEST",
  },
]
`;

exports[`feeds.actions feedsGetAll() handles getting feeds 1`] = `
Array [
  Object {
    "feeds": Immutable.OrderedMap {
      "5b33c76cb2438d5708dc197e": Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc197e",
        "creation_date": "2018-06-27T17:20:44.037Z",
        "description": null,
        "folder_id": "5b33c57eb2438d5708dc1968",
        "link": "http://highscalability.com/blog/",
        "title": "High Scalability",
        "url": "http://feeds.feedburner.com/HighScalability",
      },
    },
    "type": "FEEDS_GETALL_COMPLETE",
  },
  Object {
    "type": "FEEDS_CLEAR_UNREAD",
  },
  Object {
    "type": "ENTRIES_GET_ALL_REQUEST",
  },
]
`;

exports[`feeds.actions feedsRefreshAll() handles refreshing entries for current feeds 1`] = `
Array [
  Object {
    "entries": Immutable.OrderedMap {},
    "type": "ENTRIES_SET_ALL",
  },
  Object {
    "filterTitle": "All",
    "filteredEntries": Immutable.OrderedMap {},
    "type": "FILTER_SET_DATA",
  },
  Object {
    "type": "FEEDS_CLEAR_UNREAD",
  },
  Object {
    "type": "ENTRIES_GET_ALL_REQUEST",
  },
]
`;
